EC-84: Revise Configuration, Customers, eContent, Login and Logging (#5)
ezsystems/commerce-documentation: docs/guide/logging/logging_api.md

# Logging API

Logging in eZ Commerce is based on Monolog. There is no separate logging API definition.
For the purpose of persisting logs within a database,
the Monolog API implementation has been extended by Doctrine-based classes and an interface.

## Important service classes and interfaces of Monolog

### Psr\Log\LoggerInterface

`LoggerInterface` is always used as the service for logging. 

> Below you can find the list of the logging methods (in order of importance).

Reviewer: ```suggestion
The list below contains the logging methods (in order of importance).
```

> The last method, `log()`, should only be used if the log level of a message can only be determined at runtime.

Reviewer: ```suggestion
Use the `log()` method only if the log level of a message can only be determined at runtime.
```

### Monolog\Logger

`Monolog\Logger` implements `LoggerInterface` but, for historical reasons, provides its own logging methods (which correspond to the interface, although are named differently).

!!! note

    Although large parts of the shop rely on `Monolog\Logger` directly, it's recommended that all new implementations rely on the PSR `LoggerInterface` instead of explicitly referring to this class.

## Base classes of Doctrine-based logging

### LogRepositoryInterface

To make database logging possible, a specific Doctrine entity class and its respective repository class must exist
for every type of log records (email log, ERP log).

> `LogRepositoryInterface` must be implemented by every repository class in order to work with the Doctrine handler.

Reviewer: I would rewrite the statement to make it less ambiguous, based on which noun in the sentence the "in order to work with the Doctrine handler" phrase applies to.

#### createNewLog()

The `createNewLog()` method instantiates a new log record entity.
It takes the following parameters:

|Parameter|Description|
|---|---|

> |`$logRecord`|The record as created by Doctrine loggers|

Reviewer: ```suggestion
|`$logRecord`|The record, as created by Doctrine loggers|
```

|`$persist`|If `true`, calls `save()` implicitly|

#### save()

The `save()` method persists the given log entity.
It takes the following parameters:

|Parameter|Description|
|---|---|
|`$logEntity`|A log entity, created by `createNewLog()`|

### AbstractOrmLogRepository

> This implementation of `LogRepositoryInterface` uses doctrines ORM API by deriving from `EntityRepository`.

Reviewer: ```suggestion
This implementation of `LogRepositoryInterface` uses the Dotrine ORM API by deriving from `EntityRepository`.
```

The class is abstract, but implements the complete `LogRepositoryInterface`.
It is intended to be extended for concrete entity implementations of `AbstractLog` and, therefore, must not be instantiated directly.

It's recommended to extend this class for any ORM-based logging entities.

### AbstractLog

`AbstractLog` is the abstract class for all Log entities.

> It cannot be persisted by itself and MUST be derived by, for example, a fully-mapped doctrine entity class.

Reviewer: ```suggestion
It cannot be persisted by itself and MUST be derived by, for example, a fully-mapped Doctrine entity class.
```
